container-core/src/main/java/com/yahoo/container/jdisc/state/MetricsPacketsHandler.java

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.container.jdisc.state;
+
+import com.google.inject.Inject;
+import com.yahoo.collections.Tuple2;
+import com.yahoo.component.provider.ComponentRegistry;
+import com.yahoo.container.jdisc.state.StateHandler.JSONObjectWithLegibleException;
+import com.yahoo.jdisc.Request;
+import com.yahoo.jdisc.Response;
+import com.yahoo.jdisc.Timer;
+import com.yahoo.jdisc.handler.AbstractRequestHandler;
+import com.yahoo.jdisc.handler.ContentChannel;
+import com.yahoo.jdisc.handler.ResponseDispatch;
+import com.yahoo.jdisc.handler.ResponseHandler;
+import com.yahoo.jdisc.http.HttpHeaders;
+import com.yahoo.metrics.MetricsPresentationConfig;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.yahoo.container.jdisc.state.StateHandler.getSnapshotPreprocessor;
+
+/**
+ * This handler outputs metrics in a json-like format. Each individual metric is a json object (packet),
+ * but there is no outer array or object that wraps the metrics packets. This handler is not set up by
+ * default, but can be added to the applications's services configuration.
+ *
+ * Based on {@link StateHandler}.
+ *
+ * @author gjoranv
+ */
+public class MetricsPacketsHandler extends AbstractRequestHandler {
+    static final String APPLICATION_KEY = "application";
+    static final String TIMESTAMP_KEY   = "timestamp";
+    static final String STATUS_CODE_KEY = "status_code";
+    static final String STATUS_MSG_KEY  = "status_msg";
+    static final String METRICS_KEY     = "metrics";
+    static final String DIMENSIONS_KEY  = "dimensions";
+
+
+    private final StateMonitor monitor;
+    private final Timer timer;
+    private final SnapshotProvider snapshotPreprocessor;
+    private final String applicationName;
+
+    @Inject
+    public MetricsPacketsHandler(StateMonitor monitor,
+                                 Timer timer,
+                                 ComponentRegistry<SnapshotProvider> preprocessors,
+                                 MetricsPresentationConfig presentation,
+                                 MetricsPacketsHandlerConfig config) {
+        this.monitor = monitor;
+        this.timer = timer;
+        snapshotPreprocessor = getSnapshotPreprocessor(preprocessors, presentation);
+        applicationName = config.application();
+    }
+
+
+    @Override
+    public ContentChannel handleRequest(Request request, ResponseHandler handler) {
+        new ResponseDispatch() {
+            @Override
+            protected Response newResponse() {
+                Response response = new Response(Response.Status.OK);
```

Comment:
Return `405 Method Not Allowed` response for every method except GET.

Reply:
This handler is protocol agnostic (comment added).